---
title: "R Notebook"
output: html_notebook
---

load the necessary libraries
```{r}
library("seqinr") # is a package designed to process and analyse sequence data.
library("R.utils") # general utilities like zip and unzip
```

1. Download the whole set of coding DNA sequences for E. coli and your organism of interest. How many coding sequences are present in these organisms? Present this in the form of a table. Describe any differences between the two organisms. 

a. Downloading E.coli data

```{r}
URL="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz") # Downloads the E.coli file from the URL
gunzip("ecoli_cds.fa.gz") # Unzips the E.coli gz file
list.files() # Lists all files to check if it worked
```

b. Downloading organism of interest (C. baratii) data

```{r}
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-59/fasta/bacteria_14_collection/clostridium_baratii_gca_001405755/cds/Clostridium_baratii_gca_001405755.14207_7_79.cds.all.fa.gz"
download.file(URL,destfile="clostridium_cds.fa.gz") # Downloads the C.baratii file from the URL
gunzip("clostridium_cds.fa.gz") # Unzips the C.baratii gz file
list.files() # Lists all files to check if it worked
```

c. Reading the data for both organisms 

```{r}
#Reading the uncompressed FASTA file containing coding sequences into R for both organisms
cds_ecoli <- seqinr::read.fasta("ecoli_cds.fa") 
cds_Cbaratii <- seqinr::read.fasta("clostridium_cds.fa") 
```

d. Finding the number of coding sequences

```{r}
sequences_table <- data.frame(Organism= c("Escherichia coli", "Clostridium baratii"), Num_Coding_Sequences= c(length(cds_ecoli),length(cds_Cbaratii))) # Creates the table with the total length data
sequences_table # View the table
```

2. How much coding DNA is there in total for these two organisms?

a. E. coli

```{r}
seqinr::count(unlist(cds_ecoli),1) # Counts how many nucleotides A,T,G and C in the entirety of E. coli coding sequences
sum(seqinr::count(unlist(cds_ecoli),1)) # Calculates the total nucleotide frequency in E. coli coding sequences
```

b. C. baratii

```{r}
seqinr::count(unlist(cds_Cbaratii),1)  # Counts how many nucleotides A,T,G and C in the entirety of C. baratii coding sequences
sum(seqinr::count(unlist(cds_Cbaratii),1)) # Calculates the total nucleotide frequency in C. baratii coding sequences 
```
Create a table for the results of E.coli and C.baratii 
```{r}
Coding_DNA_Table <- data.frame(Organism= c("Escherichia coli", "Clostridium baratii"), Total_Coding_DNA= c(sum(seqinr::count(unlist(cds_ecoli),1))  ,sum(seqinr::count(unlist(cds_Cbaratii),1)))) # Defining the table for total coding sequences for each organism 
Coding_DNA_Table
```


3. Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.

```{r}
len_ecoli <- as.numeric(summary(cds_ecoli)[,1]) # Gets the length of all sequences from the first column
len_Cbaratii <- as.numeric(summary(cds_Cbaratii)[,1]) # Gets the length of all sequences from the first column

#Getting the mean and median length for each organism 
mean(len_ecoli) # Base R Mean Function
median(len_ecoli) # Base R Median Function

mean(len_Cbaratii)
median(len_Cbaratii)

#Printing the mean and median lengths
print("The mean coding sequence length of E.coli is 938.5534 and the median is 831")
print("The mean coding sequence length of C.baratii is 920.7707 and the median is 792")

```

Boxplot of the coding sequence lengths

```{r}

# Combine lengths into a single data frame for boxplot
lengths_data <- data.frame(
  Length = c(len_ecoli, len_Cbaratii),
  Organism = factor(c(rep("E.coli", length(len_ecoli)), rep("C.baratii", length(len_Cbaratii))))
)

# Now create a boxplot
boxplot(Length ~ Organism, data = lengths_data,
        main = "Boxplot of Coding Sequence Lengths",
        xlab = "Organism",
        ylab = "Length of Coding Sequences ",
        col = c("lightblue", "lightgreen"),
        border = "black",
        boxwex = 0.5
        )

mtext("Figure 1: Boxplot of total coding sequence length for C.baratii (blue) and E. coli (green)", side = 1, line = 4, cex = 0.8)
```


4. Calculate the frequency of DNA bases in the total coding sequences for both organisms. Perform the same calculation for the total protein sequence. Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.

```{r}
count(unlist(cds_ecoli),1)
count(unlist(cds_Cbaratii),1)
```


a. E. coli


```{r}
ecoli_dna_composition <- count(unlist(cds_ecoli),1)
barplot(ecoli_dna_composition,xlab="Nucleotides",ylab="Frequency", main="E.coli nucleotide composition", names.arg=toupper(names(ecoli_dna_composition)))
```

b. C. baratii

```{r}
cbaratii_dna_composition <- count(unlist(cds_Cbaratii),1)
barplot(cbaratii_dna_composition,xlab="Nucleotides",ylab="Frequency", main="C.baratii nucleotide composition",names.arg=toupper(names(ecoli_dna_composition)))
```


translation to amino acids
```{r}
prot_ecoli <- lapply(cds_ecoli,translate)
prot_cbaratii <- lapply(cds_Cbaratii,translate)
```





